### PR TITLE
Anc search by mobile

### PR DESCRIPTION
**Pull request by SebaMutuku**

### CODE REVIEW

**Review by dubdabasoduba — CHANGES_REQUESTED** [review]
Please check this branch this PR is pointing to. I don't think it should be pointing to master.

File: pom.xml [hunk 1]
```diff
@@ -17,15 +17,14 @@
         <hibernate-entitymanager.version>5.4.24.Final</hibernate-entitymanager.version>
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
         <spring.version>5.2.21.RELEASE</spring.version>
-        <spring.security.version>5.3.13.RELEASE</spring.security.version>
-        <spring.security.oauth.version>2.5.1.RELEASE</spring.security.oauth.version>
+        <spring.security.version>5.2.15.RELEASE</spring.security.version>
+        <spring.security.oauth.version>2.4.2.RELEASE</spring.security.oauth.version>
```
Comment (dubdabasoduba) [inline pom.xml:21]: Why are we downgrading?

File: pom.xml [hunk 2]
```diff
@@ -17,15 +17,14 @@
...
         <spring.fox.swagger.version>3.0.0</spring.fox.swagger.version>
         <spring.data.redis>2.2.13.RELEASE</spring.data.redis>
         <redis.jedis.version>3.8.0</redis.jedis.version>
-        <redis.lettuce.version>6.1.6.RELEASE</redis.lettuce.version>
         <opensrp.updatePolicy>always</opensrp.updatePolicy>
         <nexus-staging-maven-plugin.version>1.5.1</nexus-staging-maven-plugin.version>
-        <opensrp.core.version>3.1.1-SNAPSHOT</opensrp.core.version>
+        <opensrp.core.version>2.13.4-SNAPSHOT</opensrp.core.version>
```
Comment (dubdabasoduba) [inline pom.xml:27]: Why are we downgrading?

File: pom.xml [hunk 3]
```diff
@@ -375,14 +375,14 @@
...
-            <version>5.5.1</version>
+            <version>5.6.0</version>
         </dependency>
 
         <!-- https://mvnrepository.com/artifact/io.lettuce/lettuce-core -->
         <dependency>
             <groupId>io.lettuce</groupId>
             <artifactId>lettuce-core</artifactId>
-            <version>${redis.lettuce.version}</version>
+            <version>6.1.6.RELEASE</version>
```
Comment (dubdabasoduba) [inline pom.xml:385]: Any reason for the change to reference the version directly?

File: src/main/java/org/opensrp/web/Constants.java [hunk 4]
```diff
@@ -97,4 +98,27 @@ interface Metrics {
...
+        String TEMPLATE_NOT_FOUND = "Plan template not found";
+        String JURISDICTION_NOT_FOUND = "Jurisdiction not found";
+        String BIOPHICS_ID_MISSING = "Biophics id missing from case details event";
+        String INVALID_CASE_DETAILS_EVENT = "Case details event is invalid";
+        String MISSING_CASE_DETAILS_EVENT = "Case details event does not exist";
+        String EXTERNAL_ID = "externalId";
+        String PLAN_USER = "opensrp";
+        String BFID = "bfid";
+        int PLAN_DURATION_DAYS = 20;
+    }
```
Comment (dubdabasoduba) [inline src/main/java/org/opensrp/web/Constants.java:123]: This seems like a Reveal requirement. Any reason for it on this PR?

**Review by dubdabasoduba — CHANGES_REQUESTED** [review]

**Comment by dubdabasoduba** [comment]
@SebaMutuku this PR seems to have extra pieces of work. Please remember to make each commit and by extension each PR atomic. I.e each PR handles one issue unless the issues are related.

**Comment by SebaMutuku** [comment]
@dubdabasoduba , I had pulled code from latest tag release which v2.2 branch and checked out the above referenced branch from there. However, it seems v2.2 which has latest changes isn't upto date with master.

File: src/main/java/org/opensrp/web/rest/SearchResource.java [hunk 5]
```diff
@@ -9,10 +9,7 @@
 import static org.opensrp.web.rest.RestUtils.getStringFilter;
 
 import java.text.ParseException;
-import java.util.ArrayList;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
+import java.util.*;
```
Comment (dubdabasoduba) [inline src/main/java/org/opensrp/web/rest/SearchResource.java:12]: Please avoid importing *.

File: src/main/java/org/opensrp/web/rest/SearchResource.java [hunk 6]
```diff
@@ -60,7 +57,8 @@ public List<Client> search(HttpServletRequest request) throws ParseException {//
 		String firstName = getStringFilter(FIRST_NAME, request);
 		String middleName = getStringFilter(MIDDLE_NAME, request);
 		String lastName = getStringFilter(LAST_NAME, request);
-		
+		Optional<String> phoneNumber = Optional.ofNullable(getStringFilter("phone_number", request));
```
Comment (dubdabasoduba) [inline src/main/java/org/opensrp/web/rest/SearchResource.java:60]: Move the literal string to constants

File: src/main/java/org/opensrp/web/rest/SearchResource.java [hunk 7]
```diff
@@ -83,9 +81,13 @@ public List<Client> search(HttpServletRequest request) throws ParseException {//
 			String attributeType = StringUtils.isBlank(attributes) ? null : attributes.split(":", -1)[0];
 			String attributeValue = StringUtils.isBlank(attributes) ? null : attributes.split(":", -1)[1];
 			
-			attributeMap = new HashMap<String, String>();
+			attributeMap = new HashMap<>();
 			attributeMap.put(attributeType, attributeValue);
 		}
+		if (phoneNumber.isPresent()) {
+			attributeMap = new HashMap<>();
+			attributeMap.put("alt_phone_number", phoneNumber.get());
```
Comment (dubdabasoduba) [inline src/main/java/org/opensrp/web/rest/SearchResource.java:89]: Move the literal string to constants

File: src/test/java/org/opensrp/web/rest/it/SearchResourceTest.java [hunk 8]
```diff
@@ -133,12 +134,20 @@ public void shouldSearchClientWithGender() throws Exception {
 	public void shouldSearchClientWithBirthDate() throws Exception {
 		Client expectedClient = createOneSearchableClient();
 
-		String searchQuery = "birthdate=" + birthDate.toLocalDate().toString() + ":" + birthDate.toLocalDate().toString();
+		String searchQuery = "birthdate=" + birthDate.toLocalDate();
```
Comment (dubdabasoduba) [inline src/test/java/org/opensrp/web/rest/it/SearchResourceTest.java:137]: Any reasons we are altering this line?

**Review by dubdabasoduba — CHANGES_REQUESTED** [review]

File: src/main/java/org/opensrp/web/rest/SearchResource.java (hunk 7 above)
Comment (SebaMutuku) [reply src/main/java/org/opensrp/web/rest/SearchResource.java:89]: This means I will have to deploy a new version of server common

**Review by dubdabasoduba — CHANGES_REQUESTED** [review]
@SebaMutuku Please create a PR with the similar changes for the master branch

File: src/main/java/org/opensrp/web/rest/SearchResource.java [hunk 9]
```diff
@@ -82,18 +79,26 @@ public List<Client> search(HttpServletRequest request) throws ParseException {//
...
+		}
+		if (alternateName.isPresent()) {
+			attributeMap = new HashMap<>();
+			attributeMap.put(ALT_NAME, alternateName.get());
+		}
 		searchBean.setAttributes(attributeMap);
-		
+
 		Map<String, String> identifierMap = null;
 		String identifiers = getStringFilter("identifier", request);
```
Comment (dubdabasoduba) [inline src/main/java/org/opensrp/web/rest/SearchResource.java:97]: Please use the constants

File: src/main/java/org/opensrp/web/rest/SearchResource.java [hunk 10]
```diff
@@ -60,7 +56,8 @@ public List<Client> search(HttpServletRequest request) throws ParseException {//
 		String firstName = getStringFilter(FIRST_NAME, request);
 		String middleName = getStringFilter(MIDDLE_NAME, request);
 		String lastName = getStringFilter(LAST_NAME, request);
-		
+		Optional<String> phoneNumber = Optional.ofNullable(getStringFilter(PHONE_NUMBER, request));
+		Optional<String> alternateName = Optional.ofNullable(getStringFilter(ALT_NAME, request));
 		ClientSearchBean searchBean = new ClientSearchBean();
 		searchBean.setNameLike(getStringFilter("name", request));
```
Comment (dubdabasoduba) [inline src/main/java/org/opensrp/web/rest/SearchResource.java:62]: Please use the constants

**Review by dubdabasoduba — CHANGES_REQUESTED** [review]

**Review by dubdabasoduba — APPROVED** [review]

**Review by dubdabasoduba — CHANGES_REQUESTED** [review]
@SebaMutuku Please update the JavaDocs here too.

**Pull request merged**